source/common/upstream/cluster_discovery_manager.cc

```diff
@@ -0,0 +1,171 @@
+  }
+  ENVOY_LOG(trace, "cm cdm: invoking {} callbacks for the cluster name {} in {}",
+            callback_items.size(), name, thread_name_);
+  for (auto& item : callback_items) {
+    auto callback = std::move(item->callback_);
+    // This invalidates the handle and the invoker.
+    item.reset();
+    // The callback could be null when handle was destroyed during the
+    // previous callback.
+    if (callback != nullptr) {
+      (*callback)(cluster_status);
+    }
+  }
+}
+
+ClusterDiscoveryManager::AddedCallbackData
+ClusterDiscoveryManager::addCallback(std::string name, ClusterDiscoveryCallbackPtr callback) {
+  ENVOY_LOG(trace, "cm cdm: adding callback for the cluster name {} in {}", name, thread_name_);
+  auto& callbacks_list = pending_clusters_[name];
+  auto item_weak_ptr = addCallbackInternal(callbacks_list, std::move(callback));
+  auto handle = std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, name, item_weak_ptr);
+  CallbackInvoker invoker(*this, std::move(name), std::move(item_weak_ptr));
+  auto discovery_in_progress = (callbacks_list.size() > 1);
+  return {std::move(handle), discovery_in_progress, std::move(invoker)};
+}
+
+void ClusterDiscoveryManager::swap(ClusterDiscoveryManager& other) {
+  thread_name_.swap(other.thread_name_);
+  pending_clusters_.swap(other.pending_clusters_);
+  callbacks_.swap(other.callbacks_);
+  callbacks_handle_.swap(other.callbacks_handle_);
+}
+
+void ClusterDiscoveryManager::invokeCallbackFromItem(absl::string_view name,
+                                                     CallbackListItemWeakPtr item_weak_ptr,
+                                                     ClusterDiscoveryStatus cluster_status) {
+  auto item_ptr = item_weak_ptr.lock();
+  if (item_ptr == nullptr) {
+    ENVOY_LOG(trace, "cm cdm: not invoking an already stale callback for cluster {} in {}", name,
+              thread_name_);
+    return;
+  }
+  ENVOY_LOG(trace, "cm cdm: invoking a callback for cluster {} in {}", name, thread_name_);
+  auto callback = std::move(item_ptr->callback_);
+  if (item_ptr->self_iterator_.has_value()) {
+    eraseItem(name, std::move(item_ptr));
+  } else {
+    ENVOY_LOG(trace,
+              "cm cdm: the callback for cluster {} in {} is prepared for invoking during "
+              "processing, yet some other callback tries to invoke this callback earlier",
+              name, thread_name_);
+  }
+  if (callback != nullptr) {
+    (*callback)(cluster_status);
+  } else {
+    ENVOY_LOG(trace, "cm cdm: the callback for cluster {} in {} is prepared for invoking during "
+                     "processing, yet some other callback destroyed its handle in the meantime");
+  }
+}
+
+ClusterDiscoveryManager::CallbackList
+ClusterDiscoveryManager::extractCallbackList(absl::string_view name) {
+  auto map_node_handle = pending_clusters_.extract(name);
+  if (map_node_handle.empty()) {
+    return {};
+  }
+  CallbackList extracted;
+  map_node_handle.mapped().swap(extracted);
```

Comment:
Is swapping necessary here?

Reply:
Let me think about this case, it's been a while since I wrote it. :)